### PR TITLE
[ts] Add support for instantiation expressions

### PR DESCRIPTION
**Pull request by nicolo-ribaudo**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes https://github.com/babel/babel/issues/14322
| Patch: Bug Fix?          | 
| Major: Breaking Change?  |
| Minor: New Feature?      | Y
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/14457"></a>

### CODE REVIEW

File: packages/babel-parser/src/plugins/typescript/index.js [hunk 1]
```diff
@@ -2294,48 +2295,69 @@ export default (superClass: Class<Parser>): Class<Parser> =>
...
-            } else if (tokenIsTemplate(this.state.type)) {
-              const result = this.parseTaggedTemplateExpression(
-                base,
-                startPos,
-                startLoc,
-                state,
-              );
-              result.typeParameters = typeArguments;
-              return result;
+          if (tokenIsTemplate(this.state.type)) {
```
Comment (nicolo-ribaudo) [inline packages/babel-parser/src/plugins/typescript/index.js:2306]: This code has just been moved.

File: packages/babel-parser/src/parser/expression.js [hunk 2]
```diff
@@ -1842,6 +1842,21 @@ export default class ExpressionParser extends LValParser {
   // argument list.
   // https://tc39.es/ecma262/#prod-NewExpression
   parseNew(node: N.Expression): N.NewExpression {
+    this.parseNewCallee(node);
+
+    if (this.eat(tt.parenL)) {
```
Comment (nicolo-ribaudo) [inline packages/babel-parser/src/parser/expression.js:1847]: This code has just been moved.

**Review by existentialism — DISMISSED** [review]

File: packages/babel-generator/test/fixtures/typescript/instantiation-expression/input.ts [hunk 3]
```diff
@@ -0,0 +1,5 @@
+a<b>;
+
+(a<b>)<c>;
+(a<b>)<c>();
+new (a<b>)<c>();
```
Comment (JLHwung) [inline packages/babel-generator/test/fixtures/typescript/instantiation-expression/input.ts:5]: Can you add
```ts
(a<b>)<c>?.();
```

**Review by JLHwung — COMMENTED** [review]
- We may need to mark `TSExpressionWithTypeArguments` as an expression: Babel is throwing Property right of AssignmentExpression expected node to be of a type ["Expression"] but instead got "TSExpressionWithTypeArguments" when transpiling both TS and optional chaining ([REPL](https://babeljs.io/repl/build/51667/#?browsers=chrome%2053&build=&builtIns=false&corejs=false&spec=false&loose=false&code_lz=IYHgRgfA_AdAFASgNxA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=true&fileSize=false&timeTravel=false&sourceType=unambiguous&lineWrap=true&presets=env%2Ctypescript&prettier=false&targets=&version=7.17.9%2Bpr.14457&externalPlugins=&assumptions=%7B%22privateFieldsAsProperties%22%3Atrue%7D)):

```ts
a<b>?.();
```
Behaviour differences:
- Babel accepts `a<b><c>();` while TS rejects it with "Expression expected"
- Babel rejects `let a: typeof a<b>` while TS accepts it
- Babel parses `a>b<x>>1` as TSExpressionWithTypeArguments while TS parses them as JS.

**Comment by nicolo-ribaudo** [comment]
I just realized that `TSExpressionWithTypeArguments` is a completely different thing:
- it's "expression" can only be a TS type in the form `A.B.C`
- it's meant to be a type-only annotation

I'm tempted to update this PR to use a new node, such as `TSInstantiationExpression`, however TS uses the same node for both. (cc @bradzacher maybe you have opinions)

**Comment by sosukesuzuki** [comment]
> Babel rejects let a: typeof a<b> while TS accepts it

Regardless of whether we are adding a new node for Instantiation Expression, we should extend `TSTypeQuery` to add `typeParameters`.

**Comment by bradzacher** [comment]
> I just realized that TSExpressionWithTypeArguments is a completely different thing:
> 
> - it's "expression" can only be a TS type in the form A.B.C
> - it's meant to be a type-only annotation

are you sure this is correct?
Based on my playing around with the nightly build - ExpressionWithTypeArguments is just the node that gets used for this

```ts
declare function foo(): <T>() => T;

let a3 = foo()<string>;
//       ^^^^^^^^^^^^^ ExpressionWithTypeArguments
```

Based on TS's defined types the expression is any valid LHS expression:
https://github.com/microsoft/TypeScript/blob/56a4a93718ce016f038bcbce425bd3f47c3bec78/src/compiler/types.ts#L2473

**Comment by nicolo-ribaudo** [comment]
However, `interface A extends foo()<T> {}` (which is the place where `TSExpressionWithTypeArguments` was originally used) is not valid 🤔

**Comment by bradzacher** [comment]
Yeah TS uses the same node for all those usecase then blocks weird cases like that with a semantic error.

In our AST we have a separate node for that usecase - [`TSInterfaceHeritage`](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/ast-spec/src/special/TSInterfaceHeritage/spec.ts).

I think class implements also uses the same node and the same parsing logic within TS. Again in our AST we have a special node for this - [`TSClassImplements`](https://github.com/typescript-eslint/typescript-eslint/blob/6c3816b3831e6e683c1a7842196b34248803d69b/packages/ast-spec/src/special/TSClassImplements/spec.ts).

**Comment by nicolo-ribaudo** [comment]
Non-fun fact: in `type a = typeof import("b").c<d>` and `type e = typeof f.g<h>`, the type argument is attached in two completely different places in TS's ast: `<d>` is part of the `import` node, while `<h>` of the `typeof` node.

**Comment by nicolo-ribaudo** [comment]
So, there are multiple possible options:
1. Continue using `TSExpressionWithTypeArguments` for both interface heritage and for these new type instantiation expressions.
2. Use `TSExpressionWithTypeArguments` for both, but in Babel 8 change the one used for interface heritage to `TSInterfaceHeritage` or `TSQualifiedNameWithTypeArguments`.
3. Use `TSInstantiationExpression` for the new node, and keep `TSExpressionWithTypeArguments` for interfaces.
4. Use `TSInstantiationExpression` for the new node, and also rename `TSExpressionWithTypeArguments` in Babel 8.

My preference is 3/4 first, then 2 and then 1. I think using the same node for both (like TS does) would be particularly confusing because:
- they have completely different rules for what's valid
- one is a type, the other one is a JS expression

@JLHwung @existentialism WDYT?

**Comment by nicolo-ribaudo** [comment]
It's also possible that in Babel 8 we will be able to just use `TSTypeReference` for interfaces heritage, since it seems to have the AST shape we need.

**Comment by JLHwung** [comment]
@nicolo-ribaudo I agree with approach 4. I support renaming `TSExpressionWithTypeArguments` for interfaces since the type is confusing.

**Comment by bradzacher** [comment]
> It's also possible that in Babel 8 we will be able to just use `TSTypeReference` for interfaces heritage, since it seems to have the AST shape we need.

A consideration for why the AST is designed the way it is; it's designed to support TS's AST shape - which in turn is designed to support the "recoverable parsing" in a type-safe manner.

So the reason that we use special nodes for these is because it's another one of those "TS uses semantic errors to ban weird syntax" things.
Technically `interface Foo extends fn() {}` is parses as syntactically valid TS - though it causes TS to report a semantic error - it parses fine.
I believe that's because TS reuses its "class extends" parser logic in this location.

----

(2) would align your AST shape with the typescript-eslint estree spec as it has been defined since the beginning.
I didn't realise that babel's AST differed in this regard! [It looks like our alignment tests have taken this difference into consideration for a long while](https://github.com/typescript-eslint/typescript-eslint/blob/6c3816b3831e6e683c1a7842196b34248803d69b/packages/typescript-estree/tests/ast-alignment/utils.ts#L199-L208).

I could acquiesce to naming it `TSInstantiationExpression` to prevent naming clashes with bable's existing AST.

**Comment by nicolo-ribaudo** [comment]
@bradzacher Another reason I like `TSInstantiationExpression` is that it's name is consistent with `TSAsExpression` and `TSNonNullExpression`, which are the other two "expressions augmented with a type hint" nodes.

**Comment by nicolo-ribaudo** [comment]
Also, do you plan to represent `typeof a<b>` by adding a `.typeParameters` property to `TSTypeQuery`, or by allowing `TSTypeQuery`'s `.exprName` to be a `TSTypeReference`?

I am currently implementing it in the first way (similarly to how tsc does), but I'm annoyed because `.typeParameters` are not valid if `.exprName` is a `TSImportType` (because it already includes its own type parameters).

**Comment by bradzacher** [comment]
Interestingly I found https://github.com/babel/babel/issues/12884 which logs the difference in our ASTs for the heritage nodes!

----

> Also, do you plan to represent `typeof a<b>` by adding a `.typeParameters` property to `TSTypeQuery`, or by allowing `TSTypeQuery`'s `.exprName` to be a `TSTypeReference`?

When I saw @sosukesuzuki mention this above - I didn't hugely love the idea because it means that `typeof` further diverges from things like `keyof`.

However then I looked into it and noticed that in our AST - `typeof` maps to [`TSTypeQuery`](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/ast-spec/src/type/TSTypeQuery/spec.ts), which uses `Identifier | TSQualifiedName` for its expression - which obv doesn't support generics.

Looking into it more, for `typeof import('')`, we currently don't create a `TSTypeQuery`, instead we create a `TSImportType` with `isTypeOf: true`. We were looking to change this in our next major though (https://github.com/typescript-eslint/typescript-eslint/pull/3076).

I don't know exactly what the better option is here.
My gut is telling me that allowing the AST types to describe a potentially invalid state and then just logically banning the invalid state is acceptable for the sake of keeping things simple.

i.e. I think this is the best AST type we can define
```ts
interface TSTypeQuery {
  exprName: Identifier | TSQualifiedName | TSImportType;
  typeParameters?: TSTypeParameterInstantiation;
}
```

if you wanted you could be fancy and define a union type for it?
```ts
interface TSTypeQueryWithImport {
  exprName: TSImportType;
  typeParameters?: never;
}
interface TSTypeQueryWithName {
  exprName: Identifier | TSQualifiedName;
  typeParameters?: TSTypeParameterInstantiation;
}
type TSTypeQuery = TSTypeQueryWithImport | TSTypeQueryWithName;
```

**Comment by nicolo-ribaudo** [comment]
@bradzacher It's getting a bit out-of-scope for this PR (we might then migrate the discussion somewhere else), but since both your and our projects are considering AST breaking changes for the next major, we could make it a bit change and completely restructure the AST to something sensible like this:

```js
// typeof ...
interface TSTypeQuery {
  typeAnnotation: TSImportType | TSTypeParameterInstantiation | TSQualifiedName | Identifier;
}

// ...<x, y, z>
interface TSTypeParameterInstantiation {
  typeName: TSQualifiedName | Identifier;
  typeParameters: Array<TSType>;
}

// ... .x
interface TSQualifiedName {
  left: TSQualifiedName | TSImportType | Identifier;
  right: Identifier;
}

// import("...")
interface TSImportType {
  argument: StringLiteral;
}
```

By doing so, type parameters are only represented in `TSTypeReference` nodes, and the ASTs would look like this:
- `a.b<c>`
  ```js
  TSTypeParameterInstantiation {
    typeName: TSQualifiedName { left: Identifier("a"), right: Identifier("b") },
    typeParameters: [ "c" ]
  }
  ```
- `import("a").b<c>`
  ```js
  TSTypeParameterInstantiation {
    typeName: TSQualifiedName { left: TSImportType("a"), right: Identifier("b") },
    typeParameters: [ "c" ]
  }
  ```
- `typeof import("a").b<c>`
  ```js
  TSTypeQuery {
    typeAnnotation: TSTypeParameterInstantiation {
      typeName: TSQualifiedName { left: TSImportType("a"), right: Identifier("b") },
      typeParameters: [ "c" ]
    }
  }
  ```

This also follows the AST design for JS nodes:
- `typeof a(b)` is a `UnaryExpression(typeof)` whose argument is a `CallExpression`, and the arguments are not directly part of the outer `UnaryExpression`
- `import("a").b` is a `MemberExpression` whose object is an `ImportExpression`, not an `ImportExpression` whose "qualifier" is `b`.

**Comment by bradzacher** [comment]
> to de-dupe even further the number of places where you can find type parameters in the AST, we could also represent a<x>() and new a<x>() as a CallExpression and a NewExpression whose callee is a TSInstantiationExpression.

I don't think it's worth changing the AST completely to add the instantiation expression node in other places. It's a fairly significant change to the AST structure, and This also makes it harder to inspect the AST as it adds a layer of indirection to go through.

I.e. Much easier to do `CallExpression[type Parameters]` than it is to do `CallExpression > TSInstantiationExpression.callee` and traverse back up to the parent.

IMO when considering breaking changes to the AST, we should ensure we're making it to greatly improve the AST structure or provide some tangible wins in the representation.
I oersonally don't think that deduplicating the representation is a big enough win to justify the migration churn and pain for the ecosystem and users.

----

I agree though that cleaning up the type query representation is a big win in terms of AST structure and usability.

**Comment by nicolo-ribaudo** [comment]
TODO: https://github.com/babel/babel/pull/14457#pullrequestreview-940000258=

File: packages/babel-parser/src/plugins/typescript/index.js [hunk 4]
```diff
@@ -65,7 +65,7 @@ function assert(x: boolean): void {
   }
 }
 
-function tsTokenCanStartExpression(token: number) {
+function tsTokenCanStartExpression(token: TokenType) {
   // tsc consiers binary operators as "can start expression" tokens:
```
Comment (bradzacher) [inline packages/babel-parser/src/plugins/typescript/index.js:69]:
```suggestion
  // tsc considers binary operators as "can start expression" tokens:
```

**Review by JLHwung — APPROVED** [review]

**Pull request merged**